src/target/riscv/riscv-013.c

```diff
@@ -3396,310 +3396,524 @@ static int write_memory_abstract(struct target *target, target_addr_t address,
+		uint32_t index, uint64_t value)
+{
+	uint8_t * const buffer = access.buffer_address;
+	const uint32_t size = access.element_size;
+	const uint32_t increment = access.increment;
+	const target_addr_t address = access.target_address;
+
+	assert(size <= 8);
+	buf_set_u64(buffer + index * size, 0, 8 * size, value);
+	log_memory_access64(address + index * increment, value, size,
+			/*is_read*/ true);
+}
 
-	if (count > 1) {
-		/* Read the penultimate word. */
-		uint32_t dmi_data0, dmi_data1 = 0;
-		if (dmi_read(target, &dmi_data0, DM_DATA0) != ERROR_OK)
-			return ERROR_FAIL;
-		if (size > 4 && dmi_read(target, &dmi_data1, DM_DATA1) != ERROR_OK)
+static int read_word_from_dmi(struct target *target,
+		struct memory_access_info access, uint32_t index)
+{
+	uint32_t dmi_data0, dmi_data1 = 0;
+
+	if (dmi_read(target, &dmi_data0, DM_DATA0) != ERROR_OK)
+		return ERROR_FAIL;
+	if (access.element_size > 4) {
+		assert(access.element_size == 8);
+		assert(riscv_xlen(target) == 64);
+		if (dmi_read(target, &dmi_data1, DM_DATA1) != ERROR_OK)
 			return ERROR_FAIL;
-		uint64_t value64 = (((uint64_t)dmi_data1) << 32) | dmi_data0;
-		buf_set_u64(buffer + size * (count - 2), 0, 8 * size, value64);
-		log_memory_access64(address + size * (count - 2), value64, size, true);
 	}
 
-	/* Read the last word. */
-	uint64_t value;
-	result = register_read_direct(target, &value, GDB_REGNO_S1);
-	if (result != ERROR_OK)
-		goto error;
-	buf_set_u64(buffer + size * (count - 1), 0, 8 * size, value);
-	log_memory_access64(address + size * (count - 1), value, size, true);
+	uint64_t value = (((uint64_t)dmi_data1) << 32) | dmi_data0;
 
+	set_buffer_and_log_read(access, index, value);
 	return ERROR_OK;
+}
 
-error:
-	dmi_write(target, DM_ABSTRACTAUTO, 0);
+static int read_word_from_s1(struct target *target,
+		struct memory_access_info access, uint32_t index)
+{
+	uint64_t value;
 
-	return result;
+	if (register_read_direct(target, &value, GDB_REGNO_S1) != ERROR_OK)
+		return ERROR_FAIL;
+	set_buffer_and_log_read(access, index, value);
+	return ERROR_OK;
 }
 
-/* Only need to save/restore one GPR to read a single word, and the progbuf
- * program doesn't need to increment. */
-static int read_memory_progbuf_one(struct target *target, target_addr_t address,
-		uint32_t size, uint8_t *buffer)
+static int riscv_program_load_mprv(struct riscv_program *p, enum gdb_regno d,
+		enum gdb_regno b, int offset, unsigned int size, bool mprven)
 {
-	uint64_t mstatus = 0;
-	uint64_t mstatus_old = 0;
-	if (modify_privilege(target, &mstatus, &mstatus_old) != ERROR_OK)
+	if (mprven && riscv_program_csrrsi(p, GDB_REGNO_ZERO, CSR_DCSR_MPRVEN,
+				GDB_REGNO_DCSR) != ERROR_OK)
+		return ERROR_FAIL;
+
+	if (riscv_program_load(p, d, b, offset, size) != ERROR_OK)
 		return ERROR_FAIL;
 
-	int result = ERROR_FAIL;
+	if (mprven && riscv_program_csrrci(p, GDB_REGNO_ZERO, CSR_DCSR_MPRVEN,
+				GDB_REGNO_DCSR) != ERROR_OK)
+		return ERROR_FAIL;
+
+	return ERROR_OK;
```

Comment:
@timsifive, an aside side question: Is there a reason to handle dcsr.mprven before and after every load? Would it be possible to instead handle it at the beginning of the block read, and restore it only at the very end.

No need to change anything in this merge request, though. The change is already large enough.

Reply:
I don't think there's any reason to handle it for every memory access. In fact, it probably just could be set when the hart first halts and then not touched again (unless the user changes the setting).